content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md

```diff
@@ -0,0 +1,760 @@
+---
+layout: blog
+title: "Deploying External OpenStack Cloud Provider with Kubeadm"
+date: 2020-01-20
+slug: Deploying-External-OpenStack-Cloud-Provider-with-Kubeadm
+---
+This document describes how to install a single-master Kubernetes cluster v1.15 with kubeadm on CentOS, and then deploy an external OpenStack cloud provider and Cinder CSI plugin to use Cinder volumes as persistent volumes in Kubernetes.
+
+### Preparation in OpenStack
+
+This cluster will be running on OpenStack VMs so we'll create a few things in OpenStack for it.
+
+* A project/tenant for this Kubernetes cluster
+* a user in this project for Kubernetes, to query node information and attach volumes etc
+* a private network and subnet
+* a router for this private network and connect it to a public network for floating IPs
+* a security group for all Kubernetes VMs
+* a VM as control plane node and a few VMs as worker nodes
+
+The security group will have the following rules to open ports for Kubernetes.
+
+**Control Plane Node**
+
+|Protocol  | Port Number | Description|
+|----------|-------------|------------|
+|TCP |6443|Kubernetes API Server|
+|TCP|2379-2380|etcd server client API|
+|TCP|10250|Kubelet API|
+|TCP|10251|kube-scheduler|
+|TCP|10252|kube-controller-manager|
+|TCP|10255|Read-only Kubelet API|
+
+**Worker Nodes**
+
+|Protocol  | Port Number | Description|
+|----------|-------------|------------|
+|TCP|10250|Kubelet API|
+|TCP|10255|Read-only Kubelet API|
+|TCP|30000-32767|NodePort Services|
+
+**CNI ports on both control plane and worker nodes**
+
+|Protocol  | Port Number | Description|
+|----------|-------------|------------|
+|TCP|179|Calico BGP network|
+|TCP|9099|Calico felix (health check)|
+|UDP|8285|Flannel|
+|UDP|8472|Flannel|
+|TCP|6781-6784|Weave Net|
+|UDP|6783-6784|Weave Net|
+
+CNI specific ports are only required to be opened when that particular CNI plugin is used. In this instruction we use Weave net, thus only those Weave net ports, TCP 6781-6784 and UDP 6783-6784, need to be opened in the security group.
+
+The control plane node needs at least 2 cores and 4GB RAM. After the VM is launched, verify its hostname and make sure it is the same as the node name in Nova. 
+If the hostname is not resolvable, add it to `/etc/hosts`.
+
+For example, if the VM is called master1, and it has an internal IP 192.168.1.4. Add that to `/etc/hosts` and set hostname to master1.
```

Comment:
Please change all instances of "master" to "control plane."

Reply:
fixed